url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tf/CompositeTf.java

```diff
@@ -0,0 +1,531 @@
+/* 
+ * es.upc.tffst
+ * Copyright (C) 2004 Javier Baliosian
+ * All rights reserved.
+ *
+ */
+
+package uy.edu.fing.mina.fsa.tf;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+
+import uy.edu.fing.mina.fsa.logics.Utils;
+
+public class CompositeTf extends Tf {
+
+  /**
+   * Comment for <code>serialVersionUID</code>
+   */
+  private static final long serialVersionUID = 1L;
+
+  public TfI left;
+
+  public String op;
+
+  public TfI right;
+  
+  public CompositeTf() {
+    super();
+  }
+
+  public CompositeTf(String op, TfI leftTf, TfI rightTf) {
+    super();
+    
+    this.op = op;
+    this.left = leftTf;
+    this.right = rightTf;
+    
+  }
+
+  /*
+   * Methods
+   * **************************************************************************
+   */
+
+
+  /**
+   * @param tf
+   * @return
+   */
+  public static ArrayList<TfI> ctfToArrayList(TfI tf) {
+    ArrayList<TfI> listOfTfs = new ArrayList<TfI>();
+
+    if (tf instanceof SimpleTf) {
+      listOfTfs.add(tf);
+      return listOfTfs;
+    } else {
+      if (tf instanceof CompositeTf) {
+        CompositeTf ctf = (CompositeTf) tf;
+        listOfTfs.addAll(ctfToArrayList(ctf.left));
+        listOfTfs.addAll(ctfToArrayList(ctf.right));
+      }
+    }
+    return listOfTfs;
+  }
+
+  public Object clone() throws CloneNotSupportedException {
+    CompositeTf clon = (CompositeTf) super.clone();
+
+    if (left instanceof SimpleTf)
+      clon.left = left;
+    else
+      clon.left = (TfI) left.clone();
+    if (right instanceof SimpleTf)
+      clon.right = right;
+    else
+      clon.right = (TfI) right.clone();
+    
+    clon.op = op;
+    
+//    clon.formula = formula;
+
+    return clon;
+  }
+
+  public boolean equals(Object o) {
+	if (o instanceof CompositeTf) {
+	  CompositeTf ctf = (CompositeTf) o;
+
+	  CompositeTf simpleThis = (CompositeTf) Utils.simplify(this);
+	  CompositeTf simpleThat = (CompositeTf) Utils.simplify(ctf);
+
+	  if (simpleThis.op == simpleThat.op && simpleThis.isEpsilon() == simpleThat.isEpsilon() && simpleThis.isNot() == simpleThat.isNot()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.